File: ForexConnectAPI-1.1.2-Linux-i686/samples/java/WholeSample/src/com/fxcore2/samples/ResponseListener.java
package com.fxcore2.samples;

import java.util.concurrent.ConcurrentLinkedQueue;

import com.fxcore2.IO2GResponseListener;
import com.fxcore2.O2GResponse;

public class ResponseListener implements IO2GResponseListener{

    private SimpleLog mSimpleLog;
    private ConcurrentLinkedQueue<O2GResponse> mResponseQueue;
    private EventPipe mEventPipe;
    
    public ResponseListener(SimpleLog simpleLog, ConcurrentLinkedQueue<O2GResponse> responseQueue, EventPipe eventPipe) {
        mSimpleLog = simpleLog;
        mResponseQueue = responseQueue;
        mEventPipe = eventPipe;
    }

    @Override
    public void onRequestCompleted(String sRequestID, O2GResponse response) {
        mResponseQueue.add(response);
        mEventPipe.putEvent(EventType.RESPONSE);
    }

    @Override
    public void onRequestFailed(String sRequestID, String sError) {
        mSimpleLog.log("Request failed requestID=%s error=%s", sRequestID, sError);
    }

    @Override
    public void onTablesUpdates(O2GResponse response) {
        mSimpleLog.log("Data received");
        mResponseQueue.add(response);
        mEventPipe.putEvent(EventType.RESPONSE);
    }
}

File: ForexConnectAPIRobot/samples/cpp/SimpleSamples/NonTableManagerSamples/TwoConnection/source/ResponseListener.cpp
#include "stdafx.h"
#include "SimpleLog.h"
#include "ResponseQueue.h"
#include "ResponseListener.h"

ResponseListener::ResponseListener(SimpleLog *log,
                                   ResponseQueue *queue)
{
    mResponse = uni::CreateEvent(NULL, FALSE, FALSE, NULL);
    mQueue = queue;
    mLog = log;
    dwRef = 0;
}

ResponseListener::~ResponseListener()
{
    uni::CloseHandle(mResponse);
}
/** */
void ResponseListener::onRequestCompleted(const char * requestId, IO2GResponse  *response)
{
    response->addRef();
    mQueue->push(response);
    uni::SetEvent(mResponse);
}

void ResponseListener::onRequestFailed(const char *requestID , const char *error)
{
    char buffer[1024];
    sprintf_s(buffer, sizeof(buffer), "Request failed requestID=%s error=%s", requestID, error);
    mLog->logString(buffer);
}

void ResponseListener::onTablesUpdates(IO2GResponse *data)
{
    mLog->addStringToLog("Data recieved");
    mLog->dumpLog();
    data->addRef();
    mQueue->push(data);
    uni::SetEvent(mResponse);
}

long ResponseListener::addRef()
{
    return InterlockedIncrement(&dwRef);
}

long ResponseListener::release()
{
    LONG dwRes = InterlockedDecrement(&dwRef);
    if (dwRes == 0)
        delete this;
    return dwRes;
}

